.ci/CI/src/modules/UnitTest.jl

```
@@ -3,122 +3,195 @@ Specify target processor for the tests.
+        )
     )
 
-Add an unit test to job_dict for a given Julia version. The generated job contains all properties
-to be directly translated to GitLab CI yaml.
+Add an unit test to job_dict for a given unit test type and target test platform. The generated job
+contains all properties to be directly translated to GitLab CI yaml.
 
 # Args
 - `job_dict::Dict`: Dict in which the new job is added.
 - `test_package::TestPackage`: Properties of the package to be tested, such as name and version.
-- `julia_versions::Vector{String}`: Julia version used for the tests.
-- `target_branch::AbstractString`: A different job code is generated depending on the target branch.
+- `setup_dev_env::Bool`: If the value is true, additional job code is generated that allows the dev
+    or feature branch versions of the QED dependencies to be used.
+- `unit_test_type::UnitTestType`: Depending on the type, slightly different unit tests are generated.
+    Read the documentation of the concrete type to get more information.
 - `test_platform::TestPlatform`: Set target platform test, e.g. CPU, Nvidia GPU or AMD GPU.
 - `tools_git_repo::ToolsGitRepo`: URL and branch of the Git repository from which the CI tools are
     cloned in unit test job.
-- `nightly_base_image::AbstractString`: Name of the job base image if the Julia version is nightly.
 """
 function add_unit_test_job_yaml!(
         job_dict::Dict,
         test_package::TestPackage,
-        julia_versions::Vector{String},
-        target_branch::AbstractString,
+        setup_dev_env::Bool,
+        unit_test_type::UnitTestType,
         test_platform::TestPlatform = CPU,
         tools_git_repo::ToolsGitRepo = ToolsGitRepo(
             "https://github.com/QEDjl-project/QuantumElectrodynamics.jl.git", "dev"
-        ),
-        nightly_base_image::AbstractString = "debian:bookworm-slim",
+        )
+    )
+    error("No special implementation for the given UnitTestType: $(unit_test_type).")
```

Is this the correct way to provide the interface for the abstract parent type `UnitTestType` without having an implementation?

---

Depends. If you want to have a hard interface, which throws if something is not implemented, you should throw a more informative `MethodError`. If it is a soft interface, which can fail silently, you just add the *name* of the function without any signature, and specify the arguments in the doc-string: 

```Julia
""" 
    
    add_unit_test_job_yaml!(
        job_dict::Dict,
        test_package::TestPackage,
        julia_versions::Vector{String},
        target_branch::AbstractString,
        setup_dev_env::Bool,
        unit_test_type::UnitTestType,
        test_platform::TestPlatform = CPU,
        tools_git_repo::ToolsGitRepo = ToolsGitRepo(
            "https://github.com/QEDjl-project/QuantumElectrodynamics.jl.git",
            "dev"
        )
    )

Interface function, which needs to be concretely implemented for a given `UnitTestType`
"""
function add_unit_test_job_yaml end
```

Calling this function with some arguments will throw a `MethodError`, showing exactly which argument has nothing to dispatch on.

---

I think, the `MethodError` makes sens. I want to have a hard interface and if we maybe add more  `UnitTestType` types, it should throw an error, if there is no dedicated generator function is implemented.

---

Maybe I misunderstand something in your explanation but I overtook your suggestion and it works like I want to have it.